File: GFG POTD/021-August-2025/002-longest-subarray-with-majority-greater-than-k.java
// https://www.geeksforgeeks.org/problems/longest-subarray-with-majority-greater-than-k/1




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(N)
    public int longestSubarray(int[] arr, int k) {
        int n = arr.length;
        Map<Integer, Integer> mpp = new HashMap<>();
        int maxLen = 0, sum = 0;
        
        for (int i = 0; i < n; i ++) {
            sum += (arr[i] > k) ? 1 : -1;
           
            if (sum > 0) {
                maxLen = Math.max(maxLen, i + 1);
            } else {
                if (mpp.containsKey(sum - 1)) {
                    maxLen = Math.max(maxLen, i - mpp.get(sum - 1));
                }
            }
            
            if (!mpp.containsKey(sum)) {
                mpp.put(sum, i);
            }
        }
        return maxLen;
    }
}

File: GFG POTD/021-August-2025/002-longest-subarray-with-majority-greater-than-k.cpp
// https://www.geeksforgeeks.org/problems/longest-subarray-with-majority-greater-than-k/1




#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    // TC : O(N)
    // SC : O(N)
    int longestSubarray(vector<int> &arr, int k) {
        int n = arr.size();
        unordered_map<int, int> firstSeen;
        int prefix = 0;
        int maxLen = 0;
    
        for (int i = 0; i < n; ++i) {
            // transform
            int val = (arr[i] > k) ? 1 : -1;
            prefix += val;
    
            if (prefix > 0) {
                maxLen = i + 1;  // from 0 to i is valid
            } else {
                // if prefix - 1 seen before
                if (firstSeen.count(prefix - 1)) {
                    maxLen = max(maxLen, i - firstSeen[prefix - 1]);
                }
            }
    
            // store first occurrence only
            if (!firstSeen.count(prefix)) {
                firstSeen[prefix] = i;
            }
        }
    
        return maxLen;
    }
};